src/speciation.cpp

```diff
@@ -387,57 +394,52 @@ bool SpeciationMove::deactivateAllReactants(Change &change) {
                     faunus_logger->warn("all {} molecules have been deleted; increase system volume?",
                                         Faunus::molecules[molid].name);
                 }
-                return false;
+                throw SpeciationMoveException();
             }
         }
     }
-    return true;
 }
 
 /**
  * Checks if there is enough implicit molecules to carry out the reaction
  */
 bool SpeciationMove::enoughImplicitMolecules() const {
-    auto tooSmall = [&](auto &i) {                                // check if species has enough implicit
+    auto isExhausted = [&](auto &i) {                             // check if species has enough implicit
         auto [molid, nu] = i;                                     // molid and stoichiometric coefficient
         if (Faunus::molecules[molid].isImplicit()) {              // matter to perform process
             assert(spc.getImplicitReservoir().count(molid) == 1); // must be registered in space!
-            if (spc.getImplicitReservoir()[molid] - nu < 0) {
-                return true; // nope, no more material
-            }
+            return nu > spc.getImplicitReservoir()[molid];        // not enough material?
+        } else {
+            return false;                                         // non-implicit molecules cannot be exhausted
         }
-        return false;
     };
 
     [[maybe_unused]] auto [atomic_reactants, molecular_reactants] = reaction->getReactants();
 
-    if (std::count_if(molecular_reactants.begin(), molecular_reactants.end(), tooSmall) > 0) {
-        return false;
-    } else {
-        return true;
-    }
+    return std::find_if(molecular_reactants.begin(), molecular_reactants.end(), isExhausted) ==
+           molecular_reactants.end(); // no molecule exhausted?
 }
 
 void SpeciationMove::_move(Change &change) {
     assert(other_spc != nullptr);        // knowledge of other space should be provided by now
-    if (not Faunus::reactions.empty()) { // global list of reactions
+    try {
```

Comment:
Have you checked that all tests pass? The boolean return type is used to catch a number of things (I can't remember them all) that could influence the speciation flow. I like the idea of throwing much more so we can get rid of original comments like "slip out the back door" :-)